sale_delivery_rate/models/delivery_carrier.py

```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+from openerp import api, fields, models
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = 'delivery.carrier'
+
+    delivery_type = fields.Selection(
+        selection_add=[('base_on_rate', 'Based on Rates')],
+    )
+
+    @api.model
+    def base_on_rate_get_shipping_price_from_so(self, sale_orders):
+        prices = []
+
+        for sale_order in sale_orders:
+            order_rate = self.env['delivery.carrier.rate'].search([
```

Comment:
Good point. There's no business reason to have multiple rates for the same sale order and carrier, but the field definitions don't actually prohibit this, so adding a limit would protect against this edge case. On the other hand, it might actually be best to let this scenario generate an exception since the connector shouldn't be creating it in the first place.

@lasley - What would you recommend here?